pageserver/src/restore_local_repo.rs

```diff
@@ -440,8 +679,32 @@ fn save_xlog_smgr_truncate(timeline: &dyn Timeline, lsn: Lsn, rec: &XlSmgrTrunca
 /// Subroutine of save_decoded_record(), to handle an XLOG_XACT_* records.
 ///
 /// We are currently only interested in the dropped relations.
-fn save_xact_record(timeline: &dyn Timeline, lsn: Lsn, rec: &XlXactParsedRecord) -> Result<()> {
-    for xnode in &rec.xnodes {
+fn save_xact_record(
+    timeline: &dyn Timeline,
+    lsn: Lsn,
+    parsed: &XlXactParsedRecord,
+    decoded: &DecodedWALRecord,
+) -> Result<()> {
+    // Record update of CLOG page
+    let mut blknum = parsed.xid / pg_constants::CLOG_XACTS_PER_PAGE;
+    let tag = ObjectTag::Clog(SlruBufferTag { blknum });
+    let rec = WALRecord {
+        lsn,
+        will_init: false,
+        rec: decoded.record.clone(),
+        main_data_offset: decoded.main_data_offset as u32,
+    };
+    timeline.put_wal_record(tag, rec.clone())?;
```

Comment:
Since we're fully parsing the xlog records here anyway, it seems wasteful to keep it in the PostgreSQL WAL record format when it's stored in the repository. Because then you need to use the Postgres WAL redo instance to replay them. We should have a better way to describe modifications to clog pages. (and same with pg_multixact, relmapper etc.)

Reply:
Well, I partly agree with you. Parsing WAL record twice is really bad smell.
But we have to serialize WAL record in any case to store in object storage.
Having two different formats and so two different parsers seems to be even worse approach.